### DISCUSSION TITLE
Lifting state up to our _app.js causes problems

### DISCUSSION THREAD

**Discussion by alinarub** [neuefische/questions#765] [Web] [answered]
We successfully lifted up our assignedPoints state to the _app.js. This state saves the dogs and their points, that the user submitted in the quiz. We also lifted up the function assignPointsToDog to the _app.js. It all works great if everything is on the quiz-page. However, if we try to show the results on the results-page, the state doesn't transfer. It's an empty array.

How can we make it work, not only on the quiz-page but also on the results-page?

P.S. We also used `router.push` to jump to the results-page after the submit button is clicked. Strangely, we have to click the submit button twice in order to be shown the results array in the console.log.

Thank you in advance!

**Comment by NikRoe** [comment]
Hi Alina,

can you send us a PR with the current state of your app?

**Comment by alinarub** [comment]
Here is the [PR](https://github.com/alinarub/dog-app/pull/13). It's a bit messy because we're in the middle of working on it. :)

**Reply by NikRoe** [reply, marked as answer]
Hi Alina,

thanks for the PR!

I tried out a few things and was able to get assigned points for all fetched dogs after submitting the form. I made a few changes and will post them here.

I moved the function handleSubmit and all its dependencies to the `_app.js`.
I worked with the fetch API to get the data from the API.
I removed some of the states that were not needed.
The point are being calculated once the dogData has arrived.

I simplified the function `calculateAssignedPoints` (this has implications on the backend, see the code for `pages/api/quizResults/index.js`)

```js
// _app.js

export default function App({ Component, pageProps }) {
  const [assignedPoints, setAssignedPoints] = useState([]);
  const router = useRouter();

  function assignPointsToDog(dog) {
    setAssignedPoints((assignedPoints) => {
      const foundDog = assignedPoints.find(
        (dogItem) => dogItem.name === dog.name
      );
      if (foundDog) {
        return assignedPoints.map((dogItem) =>
          dogItem.name === dog.name
            ? { ...dogItem, points: dogItem.points + 1 }
            : dogItem
        );
      }
      // if the dog did not exist before
      return [...assignedPoints, { name: dog.name, points: 1 }];
    });
  }

  async function getDogs(data) {
    const userValues = new URLSearchParams({
      barking: data.barking,
      trainability: data.trainability,
      energy: data.energy,
    });
    try {
      const response = await fetch(`/api/quizResults?${userValues}`);

      const data = await response.json();

      calculateAssignedPoints(data);
    } catch (error) {
      console.error(error);
    }
  }

  // Algorithmus
  function calculateAssignedPoints(dogData) {
    dogData.forEach((dog) => {
      // question #1 (barking)
      assignPointsToDog(dog);
      // setAssignedPoints for these dogs +1
    });
  }

  async function handleSubmit(event) {
    event.preventDefault();
    // Collect form data
    const formData = new FormData(event.target);
    const data = Object.fromEntries(formData);

    // Get dogs from API
    await getDogs(data);

    router.push("/quiz-results");
  }

  return (
    <>
      <GlobalStyle />
      <Component
        {...pageProps}
        assignedPoints={assignedPoints}
        handleSubmit={handleSubmit}
      />
    </>
  );
}
```

Afterwards, you can remove a lot of code from your QuizPage:

```js
// quiz-page/index.js

export default function QuizPage({ handleSubmit }) {
  return (
    <div>
      <IntroText />
      <QuizForm onSubmit={handleSubmit} />
    </div>
  );
}
```

In your backend I made a small change, when you send your response you sent an array of arrays, whereas I spread those three array into a new one:

```js
// pages/api/quizResults/index.js

// ...
 // Backend sends response from external API back to the client
    response
      .status(200)
      .json([...dataBarking, ...dataEnergy, ...dataTrainability]);
    return;
// ...

```

I hope this already helps!

Let me know on Monday and have a great weekend 🙌
